core/domain/html_cleaner.py

```diff
@@ -447,32 +517,81 @@ def validate_textangular_format(html_list, run_migration=False):
         soup = bs4.BeautifulSoup(
             soup_data.replace('<br>', '<br/>'), 'html.parser')
 
-        # Text with no parent tag is also invalid.
-        for content in soup.contents:
-            if not content.name:
-                err_dict['strings'].append(html_data)
-                break
-
-        for tag in soup.findAll():
-            # Checking for tags not allowed in RTE.
-            if tag.name not in allowed_tag_list:
-                if 'invalidTags' in err_dict:
-                    err_dict['invalidTags'].append(tag.name)
-                else:
-                    err_dict['invalidTags'] = [tag.name]
-                err_dict['strings'].append(html_data)
-            # Checking for parent-child relation that are not
-            # allowed in RTE.
-            parent = tag.parent.name
-            if (tag.name in allowed_tag_list) and (
-                    parent not in allowed_parent_list[tag.name]):
-                if tag.name in err_dict:
-                    err_dict[tag.name].append(parent)
-                else:
-                    err_dict[tag.name] = [parent]
+        is_invalid = validate_soup_for_textangular(soup, err_dict)
+
+        if is_invalid:
+            err_dict['strings'].append(html_data)
+
+        for collapsible in soup.findAll('oppia-noninteractive-collapsible'):
+            # content is taken from [-1:1] since quotes are unescaped
+            # and are treated as part of html content.
+            content_html = html_unescaper(
+                collapsible['content-with-value'])[1:-1]
+            soup_for_collapsible = bs4.BeautifulSoup(
+                content_html.replace('<br>', '<br/>'), 'html.parser')
+            is_invalid = validate_soup_for_textangular(
+                soup_for_collapsible, err_dict)
+            if is_invalid:
                 err_dict['strings'].append(html_data)
 
+        for tabs in soup.findAll('oppia-noninteractive-tabs'):
+            tab_content_json = html_unescaper(tabs['tab_contents-with-value'])
+            tab_content_list = json.loads(tab_content_json)
+            for tab_content in tab_content_list:
+                content_html = tab_content['content']
+                soup_for_tabs = bs4.BeautifulSoup(
+                    content_html.replace('<br>', '<br/>'), 'html.parser')
+                is_invalid = validate_soup_for_textangular(
+                    soup_for_tabs, err_dict)
+                if is_invalid:
+                    err_dict['strings'].append(html_data)
+
     for key in err_dict:
         err_dict[key] = list(set(err_dict[key]))
 
     return err_dict
+
+
+def validate_soup_for_textangular(soup, err_dict):
```

Comment:
Is this only used by functions in this file? If so, prefix it with an underscore.

Reply:
Done!